---
title: "Differential analysis with multiple conditions and covariates"

date: "`r Sys.Date()"
author:
- Harold Pimentel
- Nicolas Bray
- PÃ¡ll Melsted
- Lior Pachter

output:
  html_document:
    toc: false
    theme: default

abstract: "This walkthrough teaches the use of sleuth for analysis of experimental designs with multiple conditions and covariates."
---

# Introduction

The purpose of this walkthrough is to demonstrate the use of __sleuth__ for analaysis of an experiment in which a "control" condition is to be compared to multiple different experimental conditions.
We illustrate this use case with data from the paper Boj _et al_., [Organoid models of human and mouse ductal pancreatic cancer](http://www.sciencedirect.com/science/article/pii/S009286741401592X), Cell, 2015.
One of the experiments the paper describes is RNA-Seq performed on syngenic mice organoids.
Specifically, RNA was extracted from murine normal (mN), pancreatic tissues that contained low-grade murine PanIN (PanIN) and pancreatic ductal organoids from multiple primary tumors (mT).

A total of 19 samples were sequenced, according to the following experimental design:

This walkthrough explains how to test for genes differential in _at least_ one of the abnormal samples (PanIN or mT), while conditioning on the sex of the mouse.

# Preliminaries

Requirements for this tutorial:

- knowledge of how to use sleuth for simple two condition experiments (see [link here])
- __kallisto__ quantified samples from GEO GSE63348 (SRA SRP049959). [Download here](https://www.dropbox.com/s/8swoeu94candjxd/Boj_results.zip?dl=0).
- the SRA runtable. [Download here](https://raw.githubusercontent.com/pachterlab/sleuth_walkthroughs/master/boj/metadata/SraRunTable.txt).


```{r}
library('sleuth')
```

The SRA runtable looks like this:
```{r}
sample_to_condition <- read.table('../metadata/SraRunTable.txt', header = TRUE,
  sep = '\t')
head(sample_to_condition)
```

The SRA table contains metadata that is not relevant and it is convenient to simplify it with

```{r}
sample_to_condition <- dplyr::select(sample_to_condition,
  sample = run, sex = Sex_s, genotype = mouse_genotype_s)
```

## Associating transcripts to genes

As described in the Bottomly _et al_. tutorial, transcripts are associated to genes with the commands

```{r}
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host = "dec2015.archive.ensembl.org")
ttg <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "transcript_version",
  "ensembl_gene_id", "external_gene_name", "description",
  "transcript_biotype"),
  mart = mart)
ttg <- dplyr::rename(ttg, target_id = ensembl_transcript_id,
  ens_gene = ensembl_gene_id, ext_gene = external_gene_name)
head(ttg)
```

## creating the sleuth object

```{r}
sample_ids <- dir(file.path('..', 'results'))
sample_to_condition <- dplyr::mutate(sample_to_condition,
  path = file.path('..', 'results', sample_ids, 'kallisto'))
```

```{r}
so <- sleuth_prep(sample_to_condition, ~sex + genotype, target_mapping = ttg,
  aggregation_column = 'ens_gene', extra_bootstrap_summary = TRUE)
```

# Basic quality control

Before beginning analysis, it is useful to examine the overall structure of the data. The PCA plot below shows the samples and verifies that samples cluster by condition:

```{r}
plot_pca(so, color_by = 'genotype')
```

Alternatively, samples can be sorted by sex:

```{r}
plot_pca(so, color_by = 'sex')
```

We see that there appears to be 1 sample outlier.

We can find this outlier by using the sample name labels in the PCA plot:

```{r}
plot_pca(so, color_by = 'genotype', text_labels = TRUE)
```

We see that it is sample `SRR1654638`.
Unfortunately, there is no information in the `sample_to_condition` table that might indicate why this is an outlier.
It is quite possible that the outlier is real, but it is also possible that it is an artifact from possible batch effects or other experimental issues.
In this simple analysis we will remove it.
In a more complete analysis, you might want to investigate why this sample is like that.

In the next step, let's remove it.

# Removing the outlier

```{r}
sample_to_condition <- dplyr::filter(sample_to_condition, sample != 'SRR1654638')
```

```{r}
so <- sleuth_prep(sample_to_condition, ~sex + genotype, target_mapping = ttg,
  aggregation_column = 'ens_gene', extra_bootstrap_summary = TRUE)
```

We see now that the PCA is much more well-behaved and the separation between genotype is much more clear:


```{r}
plot_pca(so, color_by = 'genotype', text_labels = TRUE)
```

An examination of the MA plot (see Cuffdiff walkthrough) reveals the data to be reasonable well behaved.

# Testing for differential genes

The first test we perform is to identify genes that are differently expressed in either the mT or PanIN conditions.
To do this we first specify a "full" model for sleuth. This model contains parameters that are specific to both sex and condition.
That is, sleuth models the data as:

[math formula]

To identify differential genes sleuth will compare expression values estimated according to the full model, with those of a reduced model.
In this case, the reduced model estimates parameters only according to sex.
This has the effect of modeling the data according to the assumption that expression is independent of condition.
By comparing expression estimates deirved according to the two models, sleuth is able to identify outliers that represent genes who expression can be explained much more accurately when considering condition.

The sleuth commands for performing the differential analysis are:

```{r}
so <- sleuth_fit(so)
```

# Comparison with a naive analysis

A naive analysis of the data would be to ignore the sex of each mouse, and to just test for differences among conditions.
This is done with the commands:

```{r}
so <- sleuth_fit(so, ~genotype, 'genotype')
so <- sleuth_fit(so, ~1, 'intercept')
so <- sleuth_fit(so, 'intercept', 'genotype')
naive_results <- sleuth_results(so, 'intercept:genotype', 'lrt')
```

TODO: perform comparison to number of hits and top hits

# Testing for sex effects

To understand why te naive analysis is problematic, it is helpful to examine genes that are differ according to sex while conditioning on the experimental condition.
The commands for performing such a differential analysis are:

```{r}

so <- sleuth_lrt(so, 'genotype', 'full')
sex_results <- sleuth_results(so, 'genotype:full', 'lrt')
```

Evidently, genes such as [..] which emerge as significantly different in the naive analysis, reveal difference in sex rather than condition.
